docs: add a `What is COS` doc (#262)
canonical/observability-stack: docs/explanation/overview/what-is-cos.md

---
myst:
  html_meta:
    description: "Learn what the Canonical Observability Stack (COS) is, how it works, what it does, and the flavours available."
---

# What is COS?

The Canonical Observability Stack (COS) is a highly-integrated, low-operations observability suite powered by [Juju](https://documentation.ubuntu.com/juju/3.6/) and Kubernetes. There are two flavours available: COS (sometimes referred to as COS HA) and COS Lite.

COS gathers, processes, visualises, and alerts on telemetry — metrics, logs, and traces — generated by workloads running both within and outside of Juju. By leveraging Juju's topology model to contextualise data and charm relations to automate configuration and integration, it provides a turn-key observability solution built on best-in-class, open-source tools.

> COS is developed from the lessons learned with its predecessor, [LMA](../../../how-to/migrate/migrate-lma-to-cos-lite), and is designed to deliver a consistent, cohesive operational experience for Site Reliability Engineers.

Reviewer: ```suggestion
COS is developed from the lessons learned with its predecessor, [LMA](/how-to/migrate/migrate-lma-to-cos-lite), and is designed to deliver a consistent, cohesive operational experience for Site Reliability Engineers.
```

---
myst:
  html_meta:
    description: "Learn what the Canonical Observability Stack (COS) is, how it works, what it does, and the flavours available."
---

# What is COS?

> The **Canonical Observability Stack** (COS) is a highly-integrated, low-operations observability suite powered by [Juju](https://documentation.ubuntu.com/juju/3.6/) and Kubernetes.

Reviewer 2: ```suggestion
The Canonical Observability Stack (COS) is a highly-integrated, low-operations observability suite powered by [Juju](https://documentation.ubuntu.com/juju/3.6/) and Kubernetes. There are two editions available: COS and COS Lite.
```
This should be clarified somewhere in the intro section. I like here/upfront better, but it could also be as a new line at the bottom.

The reason I'm suggesting this is because "COS" can be used to be either COS or "COS and COS Lite" (i.e., the entire product/offering), so it's good to set that clarification early - basically implying that "COS" can refer to both editions

(I made a similar call-out in a [What is Landscape?](https://documentation.ubuntu.com/landscape/what-is-landscape/#how-landscape-works) doc)

Author: Is this okay:
```suggestion
The Canonical Observability Stack (COS) is a highly-integrated, low-operations observability suite powered by [Juju](https://documentation.ubuntu.com/juju/3.6/) and Kubernetes. There are two editions available: COS HA (more commonly referred to as COS) and COS Lite.
```

Reviewer: COS (sometimes referred to as COS HA)

COS gathers, processes, visualises, and alerts on telemetry — metrics, logs, and traces — generated by workloads running both within and outside of Juju. By leveraging Juju's topology model to contextualise data and charm relations to automate configuration and integration, it provides a turn-key observability solution built on best-in-class, open-source tools.

COS is developed from the lessons learned with its predecessor, [LMA](../../../how-to/migrate/migrate-lma-to-cos-lite), and is designed to deliver a consistent, cohesive operational experience for Site Reliability Engineers.

## How COS works

COS is deployed and operated through [Juju](https://documentation.ubuntu.com/juju/3.6/). Its components are charmed operators connected by Juju relations, which automate configuration and integration between them.

> Telemetry is collected by **aggregator charms** — such as Grafana Agent or the OpenTelemetry Collector — that run alongside the workloads being observed. These aggregators scrape or receive telemetry from their co-located workloads, then push it to the COS backends over Traefik-managed ingress endpoints. This push-based model means that the COS stack does not need network access to the observed workloads; only the aggregators need to reach the COS endpoints.

Reviewer 2: ```suggestion
Telemetry is collected by **aggregator charms**, such as Grafana Agent or the OpenTelemetry Collector, that run alongside the workloads being observed. These aggregators scrape or receive telemetry from their co-located workloads, then push it to the COS backends over Traefik-managed ingress endpoints. This push-based model means that the COS stack does not need network access to the observed workloads; only the aggregators need to reach the COS endpoints.
```

Within the stack:

- **Metrics** are ingested by Prometheus (COS Lite) or Mimir (COS).
- **Logs** are ingested by Loki.
- **Traces** are ingested by Tempo (COS only).
- Alert rules are evaluated by Prometheus and Loki, and notifications are routed through **Alertmanager**.

> - All telemetry is visualised through **Grafana**, which is pre-configured to query each backend as a datasource.

Reviewer 2: ```suggestion
- **Metrics** are ingested by Prometheus (COS Lite) or Mimir (COS).
- **Logs** are ingested by Loki.
- **Traces** are ingested by Tempo (COS only, not available in COS Lite).
- **Alert rules** are evaluated by Prometheus and Loki, and notifications are routed through Alertmanager.
- All **telemetry** is visualized through Grafana, which is pre-configured to query each backend as a datasource.
```
To make the bolding more consistent

Also, I added the "not available in COS Lite" just to make it more clear, since this is the first instance of that framing

> Juju topology labels are automatically applied to all telemetry, making it possible to filter and correlate data by model, application, unit, or charm.

Reviewer 2: Link to the Juju telemetry labels page in the COS docs

For more detail, see [Telemetry Flow](../architecture/telemetry-flow) and [Model Topology for COS Lite](../architecture/cos-lite-model-topology).

## What COS does

- **Collects telemetry**: Metrics, logs, and (in COS) traces are collected from Juju-managed workloads and, via cross-model relations, from workloads outside of Juju.
- **Stores telemetry**: Telemetry is persisted in the appropriate backend (Prometheus, Mimir, Loki, or Tempo), backed by persistent storage.

> - **Alerts**: Alert rules — shipped with charms or added by operators — are evaluated continuously. Firing alerts are routed to the configured notification channels via Alertmanager.

Reviewer 2: Is "shipped with charms or added by operators" needed? I didn't understand it - what is the difference between "charms" and "operators" here? Is this referring to non-charmed apps? (if it's not really needed, you could just remove this info)

Author: Oh right, this does sound confusing. The "operator" in this case is the Juju admin who adds alerts through COS Config. I'll update the wording so it's less confusing. How does the following sound:
```suggestion
- **Alerts**: Alert rules — whether bundled into charms or provided through the COS Configuration charm — are evaluated continuously. Firing alerts are routed to the configured notification channels via Alertmanager.
```

- **Visualises**: Grafana provides pre-built dashboards for all instrumented charms, and can be extended with custom dashboards.
- **Self-monitors**: COS collects its own telemetry, so the health of the observability stack itself is observable.
- **Scales**: COS is built on horizontally-scalable, microservices-based backends (Mimir, Loki, Tempo) for enterprise workloads.

> ## Flavours of COS

Author: I wonder if we should use the word "flavour"? We could alternatively use "edition"? "Flavour" could also refer to COS HA monolithic vs COS HA distributed so I'm wondering what word would be better to use.

Reviewer: Variant is an option too. I think "edition" doesn't work here though.
- Ubuntu "[flavors](https://ubuntu.com/desktop/flavors)"
- Fedora "[spins](https://fedoraproject.org/spins/)".
- Grafana Cloud "[tiers](https://grafana.com/pricing/)", but it's more appropriate for pricing.

@YanisaHS?

Reviewer 2: I prefer "edition" over "flavor"

Reviewer 2: Variant works!

> ## Flavours of COS

Reviewer 2: ```suggestion
## Editions of COS: COS and COS Lite
```
Thoughts on this?

Author: I'll change to "editions" for now, as you have suggested. We can discuss more in real time with the team.

Reviewer 2: I'm fine with "variant" too - see Leon's comment above.

> Two flavours are available, suited to different deployment scenarios:

Reviewer 2: ```suggestion
There are two editions available: COS and COS Lite. Each are suited to different deployment scenarios:
```
IMO it's easy to gloss over COS and COS Lite in the table header, so it should be more explicit here.

I've also suggested it in the header though. It doesn't need to be as explicit if you accept that suggestion instead

Author: I've updated the header to your suggestion, but for the sake being ultra clear, I'll apply this suggestion as well.

| Metrics backend          | Mimir (distributed)                                          | Prometheus (monolithic)                                   |
| Logs backend             | Loki (distributed)                                           | Loki (monolithic)                                         |
| Traces backend           | Tempo (distributed)                                          | Not included                                              |
| Storage                  | S3 (managed independently)                                   | PVCs, e.g. `ceph-csi` (managed independently)            |
| Resiliency               | Scalable microservices with node anti-affinity (HA-ready)    | Multi-node non-identical replication                      |
| Self-monitoring          | Metrics, logs, and traces via OpenTelemetry Collector        | Metrics only, via direct relations                        |
| Minimum requirements     | 1× 8 CPU / 16 GB RAM + storage nodes                        | 1× 4 CPU / 8 GB RAM (+ storage nodes, if any)            |

> The key architectural difference is that COS (sometimes called COS HA) is built around a **coordinator/worker pattern**: each backend (Mimir, Loki, Tempo) is split into a coordinator charm and one or more worker charms, allowing individual components to be scaled out independently and placed on separate nodes. An Nginx layer handles load balancing across workers before traffic reaches Traefik. This makes COS HA suitable for high-availability, enterprise deployments where telemetry volumes are large and resilience to node failure is required. COS Lite, by contrast, runs each backend as a single monolithic charm — simpler to deploy and far lighter on resources, but without the horizontal scalability or trace support. Both flavours share the same Grafana, Alertmanager, Traefik, and Catalogue charms.

Reviewer 2: ```suggestion
### Architecture

The key architectural difference between COS and COS Lite is how the backends are deployed.

COS (sometimes called COS HA) is built around a **coordinator/worker pattern**: each backend (Mimir, Loki, Tempo) is split into a coordinator charm and one or more worker charms, allowing individual components to be scaled out independently and placed on separate nodes. An Nginx layer handles load balancing across workers before traffic reaches Traefik. This makes COS HA suitable for high-availability, enterprise deployments where telemetry volumes are large and resilience to node failure is required. 

COS Lite, by contrast, runs each backend as a single monolithic charm, which is simpler to deploy and much lighter on resources, but without the horizontal scalability or trace support. 

Both editions share the same Grafana, Alertmanager, Traefik, and Catalogue charms.
```
This information warrants its own section

---

# What is COS?

The Canonical Observability Stack (COS) is a highly-integrated, low-operations observability suite powered by [Juju](https://documentation.ubuntu.com/juju/3.6/) and Kubernetes. There are two flavours available: COS (sometimes referred to as COS HA) and COS Lite.

COS gathers, processes, visualises, and alerts on telemetry — metrics, logs, and traces — generated by workloads running both within and outside of Juju. By leveraging Juju's topology model to contextualise data and charm relations to automate configuration and integration, it provides a turn-key observability solution built on best-in-class, open-source tools.

COS is developed from the lessons learned with its predecessor, [LMA](../../../how-to/migrate/migrate-lma-to-cos-lite), and is designed to deliver a consistent, cohesive operational experience for Site Reliability Engineers.

## How COS works

COS is deployed and operated through [Juju](https://documentation.ubuntu.com/juju/3.6/). Its components are charmed operators connected by Juju relations, which automate configuration and integration between them.

> Telemetry is collected by **aggregator charms**, such as Grafana Agent or the OpenTelemetry Collector, that run alongside the workloads being observed. These aggregators scrape or receive telemetry from their co-located workloads, then push it to the COS backends over Traefik-managed ingress endpoints. This push-based model means that the COS stack does not need network access to the observed workloads; only the aggregators need to reach the COS endpoints.

Reviewer: - Need to rephrase e.g. "collected by the OpenTelemetry Collector (replacing Grafana Agent)".
- "Load-balanced" instead of "Traefik"? Or traefik in parenthesis?

Author: For the moment I've changed this paragraph to:
```suggestion
Telemetry is collected by the OpenTelemetry Collector (replacing Grafana Agent), which runs alongside the workloads being observed. OpenTelemetry Collector scrapes or receives telemetry from its co-located workloads, then pushes it to the COS backends over ingress endpoints provided and load-balanced using Traefik. This push-based model means that the COS stack does not need network access to the observed workloads; only OpenTelemetry Collector needs to reach the COS endpoints.```
WDYT?

Within the stack:

- **Metrics** are ingested by Prometheus (COS Lite) or Mimir (COS).
- **Logs** are ingested by Loki.

> - **Traces** are ingested by Tempo (COS only, not available in COS Lite).

Reviewer: This is a replication of the telemetry overview (under explanation).
Should we link to the telemetry doc, and here focus on the delta that charmed, model driven o11y brings?

Also the `**` is an LLM giveaway :D

Author: I've changed the bullet points to:
"""
By modelling observability as a set of Juju relations, COS eliminates the manual configuration burden typically associated with spinning up a monitoring stack. Dashboards, alert rules, and scrape targets are automatically provisioned when charms are related. This application of Juju topology also means that telemetry is contextualised out of the box, enabling opadminserators to filter and correlate data by model, application, or unit without any extra instrumentation. The result is a full-stack, self-monitoring observability platform that evolves alongside the applications it observes.
"""

> - **Alert rules** are evaluated by Prometheus and Loki, and notifications are routed through Alertmanager.

Reviewer: ```suggestion
- **Alert rules** are evaluated by Prometheus, Mimir and Loki, and notifications are routed through Alertmanager.
```

- All **telemetry** is visualized through Grafana, which is pre-configured to query each backend as a datasource.

Juju topology labels are automatically applied to all telemetry, making it possible to filter and correlate data by model, application, unit, or charm. Refer to the [Juju Topology](../architecture/juju-topology) guide for more information about how Juju context is applied to telemetry.

For more detail, see [Telemetry Flow](../architecture/telemetry-flow) and [Model Topology for COS Lite](../architecture/cos-lite-model-topology).

## What COS does

> - **Collects telemetry**: Metrics, logs, and (in COS) traces are collected from Juju-managed workloads and, via cross-model relations, from workloads outside of Juju.

Reviewer: ```suggestion
- **Receives telemetry**: Metrics, logs, and (in COS) traces are received from Juju-managed workloads and, via tailored integrations, from workloads outside of Juju.
```

- **Stores telemetry**: Telemetry is persisted in the appropriate backend (Prometheus, Mimir, Loki, or Tempo), backed by persistent storage.
- **Alerts**: Alert rules — whether bundled into charms or provided through the COS Configuration charm — are evaluated continuously. Firing alerts are routed to the configured notification channels via Alertmanager.
- **Visualises**: Grafana provides pre-built dashboards for all instrumented charms, and can be extended with custom dashboards.
- **Self-monitors**: COS collects its own telemetry, so the health of the observability stack itself is observable.

> - **Scales**: COS is built on horizontally-scalable, microservices-based backends (Mimir, Loki, Tempo) for enterprise workloads.

Reviewer: Should we restructure this into a couple small paragraphs? Bullets with boldface are getting old 😅

Author: Changed to
```suggestion
COS receives metrics, logs, and (in COS) traces from workloads operated by Juju charms and, via tailored integrations, from workloads outside of Juju. Telemetry is persisted in the appropriate backends, which are backed by persistent storage.

Alert rules, whether bundled with charms or provided via the COS Configuration charm, are continuously evaluated. Firing alerts are routed through Alertmanager to configured notification channels. Grafana visualizes telemetry in its UI, displaying dashboards from related charms or the COS Configuration charm.

COS also collects its own telemetry, making the health of the observability stack observable. Bundled alert rules help monitor and validate the health of COS components and the reachability of its observed workloads.
```

## Flavours of COS: COS and COS Lite

There are two flavours available: COS and COS Lite. Each is suited to different deployment scenarios:

> |                          | **COS**                                                      | **COS Lite**                                              |

Reviewer: Table headers are auto bold, no need for the `**` here.

| Minimum requirements     | 1× 8 CPU / 16 GB RAM + storage nodes                        | 1× 4 CPU / 8 GB RAM (+ storage nodes, if any)            |

### Architecture
The key architectural difference between COS and COS Lite is how the backends are deployed.
COS (sometimes called COS HA) is built around a **coordinator/worker pattern**: each backend (Mimir, Loki, Tempo) is split into a coordinator charm and one or more worker charms, allowing individual components to be scaled out independently and placed on separate nodes. An Nginx layer handles load balancing across workers before traffic reaches Traefik. This makes COS HA suitable for high-availability, enterprise deployments where telemetry volumes are large and resilience to node failure is required. 
COS Lite, by contrast, runs each backend as a single monolithic charm, which is simpler to deploy and much lighter on resources, but without the horizontal scalability or trace support. 
Both flavours share the same Grafana, Alertmanager, Traefik, and Catalogue charms.

## Useful links

> - [COS components](../../reference/cos-components): Full list of charms, rocks, and snaps in the stack

Reviewer: It's less fragile if we use absolute links in this case,

```suggestion
- [COS components](/reference/cos-components): Full list of charms, rocks, and snaps in the stack
```

vscode won't autocomplete but meh :D
